### PR TITLE
PAGE validator: respect reading/textline order/direction

### PR DESCRIPTION
**Pull request by bertsky**
(in multi-level text consistency checks)

This takes existing ideas from `ocrd_tesserocr.recognize.page_update_higher_textequiv_levels()` into the validator (only slightly more general).

This should be considered a (late) follow-up on https://github.com/OCR-D/assets/issues/16. In particular, one should verify the refined rules look better on real-world GT which contains (concatenated) subregions, non-default `@textLineOrder` and non-default `@readingDirection`, and then re-visit the existing textual consistency rules laid out in the spec.

### CODE REVIEW

**Comment by wrznr** [comment]
# [Codecov](https://codecov.io/gh/OCR-D/core/pull/442?src=pr&el=h1) Report
> Merging [#442](https://codecov.io/gh/OCR-D/core/pull/442?src=pr&el=desc) into [master](https://codecov.io/gh/OCR-D/core/commit/d7f715fe44540058c68e02ba29e276a39b8cf67e&el=desc) will **decrease** coverage by `0.51%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/OCR-D/core/pull/442?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master     #442      +/-   ##
==========================================
- Coverage   81.93%   81.41%   -0.52%     
==========================================
  Files          39       39              
  Lines        2314     2357      +43     
  Branches      426      446      +20     
==========================================
+ Hits         1896     1919      +23     
- Misses        346      353       +7     
- Partials       72       85      +13     
```



------

[Continue to review full report at Codecov](https://codecov.io/gh/OCR-D/core/pull/442?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/OCR-D/core/pull/442?src=pr&el=footer). Last update [d7f715f...81f0133](https://codecov.io/gh/OCR-D/core/pull/442?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

**Comment by bertsky** [comment]
@kba the last commit prevents an exception (instead of a negative result) during validation when coords have less than 4 points or baselines less than 2.

> Looks great insofar as I can follow. I would really appreciate tests. We have some samples with ReadingOrder but they are very basic. Can we provide better test data? @tboenig

IIUC the main obstacle for GT activities is the lack of https://github.com/OCR-D/ocrd_segment/issues/32 which itself is due to missing progress on https://github.com/OCR-D/core/pull/443. In other words: We need a way to fix most of the gazillions of GT errors now reported by the existing (already merged) coordinate validator automatically before we can move on.

File: ocrd_validators/ocrd_validators/page_validator.py [hunk 1]
```diff
@@ -169,40 +252,45 @@ def validate_consistency(node, page_textequiv_consistency, page_textequiv_strate
...
+                textLineOrder = getattr(node, getterLO)()
+            if getterRD:
+                readingDirection = getattr(node, getterRD)()
     for class_, getter, concatenate_with in _HIERARCHY:
         if not isinstance(node, class_):
             continue
         children = getattr(node, getter)()
+        if (getter == 'get_TextRegion' and children and
+            all(child.id in readingOrder for child in children) and
+            isinstance(readingOrder[children[0].id], (OrderedGroupType, OrderedGroupIndexedType))):
```
Comment (bertsky) [inline ocrd_validators/ocrd_validators/page_validator.py:276]: Not sure I understand what I intended to do here anymore.

I now think this should read:
```suggestion
            node.id in readingOrder and
            isinstance(readingOrder[node.id], (OrderedGroupType, OrderedGroupIndexedType))):
```

**Comment by kba** [comment]
Looks good and https://github.com/PRImA-Research-Lab/PAGE-XML/issues/23 shows the benefits. I will merge tomorrow unless @tboenig objects.

**Comment by kba** [comment]
@bertsky say the word and I'll merge :)

**Comment by bertsky** [comment]
> say the word and I'll merge :)

I'd say it's not perfect now, esp. https://github.com/OCR-D/core/pull/442/commits/6bf98d0350a6fa383fe5ab6592f470f917748c0b if https://github.com/PRImA-Research-Lab/PAGE-XML/issues/23 does turn out to reveal PAGE coordinates are all "pixel-center" not "pixel-below-right" (in which case we would have **a lot** of work going over all processors) – but much better than before, anyway. So let's revisit when a verdict on coordinate semantics has been found.

**Pull request merged**